File: frontend/src/shared/api/modules.ts
import { apiClient } from "./client";

export interface ModuleRead {
  id: string;
  code: string;
  name: string;
  description: string | null;
  category: string | null;
  icon: string | null;
  color: string | null;
  is_base: boolean;
  is_active: boolean;
  current_version: string | null;
  dependencies: string[];
  installed: boolean;
  installed_version: string | null;
  created_at: string;
}

export interface AppModuleRead {
  app_id: string;
  module_id: string;
  module_code: string;
  module_name: string;
  version: string;
  status: string;
  installed_at: string;
  installed_by: string | null;
}

export interface ModuleInstallResult {
  module: ModuleRead;
  installed_dependencies: string[];
  entities_created: number;
  fields_created: number;
  pages_created: number;
}

export async function listModules(appId?: string): Promise<ModuleRead[]> {
  const { data } = await apiClient.get<ModuleRead[]>("/modules", {
    params: appId ? { app_id: appId } : undefined,
  });
  return data;
}

export async function listAppModules(appId: string): Promise<AppModuleRead[]> {
  const { data } = await apiClient.get<AppModuleRead[]>(`/apps/${appId}/modules`);
  return data;
}

export async function installModule(appId: string, moduleCode: string): Promise<ModuleInstallResult> {
  const { data } = await apiClient.post<ModuleInstallResult>(
    `/apps/${appId}/modules/${moduleCode}/install`,
  );
  return data;
}

export async function uninstallModule(appId: string, moduleCode: string): Promise<void> {
  await apiClient.delete(`/apps/${appId}/modules/${moduleCode}`);
}

File: backend/app/schemas/modules.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class ModuleVersionRead(BaseModel):
    id: uuid.UUID
    version: str
    manifest: dict
    changelog: str | None
    is_current: bool
    released_at: datetime
    model_config = {"from_attributes": True}


class ModuleRead(BaseModel):
    id: uuid.UUID
    code: str
    name: str
    description: str | None
    category: str | None
    icon: str | None
    color: str | None
    is_base: bool
    is_active: bool
    current_version: str | None = None
    dependencies: list[str] = Field(default_factory=list)
    installed: bool = False
    installed_version: str | None = None
    created_at: datetime
    model_config = {"from_attributes": True}


class AppModuleRead(BaseModel):
    app_id: uuid.UUID
    module_id: uuid.UUID
    module_code: str
    module_name: str
    version: str
    status: str
    installed_at: datetime
    installed_by: uuid.UUID | None


class ModuleConflict(BaseModel):
    kind: Literal["entity", "field", "page"]
    name: str
    entity: str | None = None   # parent entity slug for field conflicts
    source: str | None = None   # module code that owns it, or "manual"
    action: Literal["reused", "skipped"]


class ModuleInstallResult(BaseModel):
    module: ModuleRead
    installed_dependencies: list[str] = Field(default_factory=list)
    entities_created: int = 0
    fields_created: int = 0
    pages_created: int = 0
    conflicts: list[ModuleConflict] = Field(default_factory=list)
